src/StaticArraysCore.jl

```diff
@@ -491,6 +491,6 @@ Size(a::T) where {T<:AbstractArray} = Size(T)
 Size(::Type{SA}) where {SA <: StaticArray} = missing_size_error(SA)
 Size(::Type{SA}) where {SA <: StaticArray{S}} where {S<:Tuple} = @isdefined(S) ? Size(S) : missing_size_error(SA)
 
-Base.@pure Size(::Type{<:AbstractArray{<:Any, N}}) where {N} = Size(ntuple(_ -> Dynamic(), N))
+Base.@pure Size(::Type{<:AbstractArray{<:Any, N}}) where {N} = Size(ntuple(_ -> Dynamic(), Val(N)))
```

Comment:
```suggestion
Size(::Type{<:AbstractArray{<:Any, N}}) where {N} = Size(ntuple(_ -> Dynamic(), Val(N)))
```

Check `Base.assume_effects` but with this `Val` we shouldn't need `@pure`.

Reply:
Agree. Actually, removing `@pure`s was what got me touching this code initially.

The reason I didn't remove `@pure` was simply to keep things tidy, I prefer to keep PRs as small as possible so they could be reviewed more easily. Also notice that there are already open PRs for removing `@pure` for quite some time as, sadly, the maintainers are not vigilant enough in removing these dangerous macros.

Reply:
Yes, this `@pure` shouldn't be needed.

> Also notice that there are already open PRs for removing `@pure` for quite some time as, sadly, the maintainers are not vigilant enough in removing these dangerous macros.

No one ever reported a problem caused by this usage of `@pure` and at least some of them were added because they improved performance. Carefully checking if they still are not needed is at least a day of work and if we remove one that actually is still needed, there will be angry users claiming StaticArrays is not properly maintained.  So far no one checked if they actually are needed or not.